File: packages/server/src/lib/fileLaunch.ts
import { existsSync } from 'node:fs';
import { spawn, spawnSync } from 'node:child_process';
import path from 'node:path';

export interface LaunchCommand {
  readonly command: string;
  readonly args: readonly string[];
  readonly shell?: boolean;
}

export function openFileCommandFor(
  absolutePath: string,
  line?: number,
  env: Record<string, string | undefined> = process.env,
  platform: NodeJS.Platform = process.platform,
): LaunchCommand {
  const editor = resolveEditorCommand(env);
  if (editor !== undefined) {
    const target = supportsLineTarget(editor) && line !== undefined
      ? `${absolutePath}:${line}`
      : absolutePath;
    // Parse the editor into binary + args to avoid shell injection from
    // uncontrolled environment variable values (KIMI_CODE_EDITOR / VISUAL / EDITOR).
    // If the editor is a simple single binary, pass it directly to spawn with
    // the target as an argument. For editors with embedded spaces or flags
    // (e.g. 'code --goto'), split into argv. Escaped paths are rare for
    // editor env vars; fall back to platform default if parsing is ambiguous.
    const editorBin = editor.trim().split(/\s+/)[0];
    const editorArgs = editor.trim().split(/\s+/).slice(1);
    if (editorBin !== undefined && editorBin.length > 0) {
      return { command: editorBin, args: [...editorArgs, target] };
    }
    // Empty editor string — fall through to platform default.
  }

  switch (platform) {
    case 'darwin':
      return { command: 'open', args: [absolutePath] };
    case 'win32':
      return { command: 'cmd', args: ['/c', 'start', '""', absolutePath] };
    default:
      return { command: 'xdg-open', args: [absolutePath] };
  }
}

export function revealFileCommandFor(
  absolutePath: string,
  platform: NodeJS.Platform = process.platform,
): LaunchCommand {
  switch (platform) {
    case 'darwin':
      return { command: 'open', args: ['-R', absolutePath] };
    case 'win32':
      return { command: 'explorer.exe', args: [`/select,${absolutePath}`] };
    default:
      return { command: 'xdg-open', args: [path.dirname(absolutePath)] };
  }
}

export type OpenInAppId =
  | 'finder'
  | 'cursor'
  | 'vscode'
  | 'iterm'
  | 'terminal';

export const OPEN_IN_APP_IDS: readonly OpenInAppId[] = [
  'finder',
  'cursor',
  'vscode',
  'iterm',
  'terminal',
];

export interface OpenInAppOptions {
  readonly line?: number;
  readonly isDirectory?: boolean;
}

export function openInAppCommandFor(
  appId: OpenInAppId,
  absolutePath: string,
  options: OpenInAppOptions = {},
  platform: NodeJS.Platform = process.platform,
): LaunchCommand {
  switch (appId) {
    case 'vscode':
      return openInVsCodeLike('code', absolutePath, options.line, platform);
    case 'cursor':
      return openInVsCodeLike('cursor', absolutePath, options.line, platform);
    case 'finder':
      return openInFinder(absolutePath, options.isDirectory, platform);
    case 'iterm':
      return openInMacApp('iTerm', absolutePath, platform);
    case 'terminal':
      return openInMacApp('Terminal', absolutePath, platform);
  }
}

export function getAvailableOpenInApps(
  platform: NodeJS.Platform = process.platform,
): readonly OpenInAppId[] {
  return OPEN_IN_APP_IDS.filter((appId) => isOpenInAppAvailable(appId, platform));
}

function isOpenInAppAvailable(
  appId: OpenInAppId,
  platform: NodeJS.Platform,
): boolean {
  switch (appId) {
    case 'finder':
    case 'terminal':
      return platform === 'darwin';
    case 'iterm':
      if (platform !== 'darwin') return false;
      return (
        existsSync('/Applications/iTerm.app') ||
        existsSync(`${process.env['HOME'] ?? ''}/Applications/iTerm.app`)
      );
    case 'vscode':
      return commandExists('code', platform);
    case 'cursor':
      return commandExists('cursor', platform);
  }
}

function commandExists(command: string, platform: NodeJS.Platform): boolean {
  try {
    if (platform === 'win32') {
      const result = spawnSync('cmd', ['/c', 'where', command], {
        stdio: 'ignore',
      });
      return result.status === 0;
    }
    const result = spawnSync('command', ['-v', command], {
      stdio: 'ignore',
      shell: true,
    });
    return result.status === 0;
  } catch {
    return false;
  }
}

function openInVsCodeLike(
  binary: string,
  absolutePath: string,
  line: number | undefined,
  _platform: NodeJS.Platform,
): LaunchCommand {
  const target = line !== undefined ? `${absolutePath}:${line}` : absolutePath;
  const args = line !== undefined ? ['-g', target] : [target];
  return { command: binary, args };
}

function openInFinder(
  absolutePath: string,
  isDirectory: boolean | undefined,
  platform: NodeJS.Platform,
): LaunchCommand {
  switch (platform) {
    case 'darwin':
      return isDirectory
        ? { command: 'open', args: [absolutePath] }
        : { command: 'open', args: ['-R', absolutePath] };
    case 'win32':
      return isDirectory
        ? { command: 'explorer.exe', args: [absolutePath] }
        : { command: 'explorer.exe', args: [`/select,${absolutePath}`] };
    default:
      return {
        command: 'xdg-open',
        args: [isDirectory ? absolutePath : path.dirname(absolutePath)],
      };
  }
}

function openInMacApp(
  appName: string,
  absolutePath: string,
  platform: NodeJS.Platform,
): LaunchCommand {
  if (platform === 'darwin') {
    return { command: 'open', args: ['-a', appName, absolutePath] };
  }
  // These apps are macOS-only in the UI; fall back to the platform default.
  return openFileCommandFor(absolutePath, undefined, process.env, platform);
}

export async function launchDetached(cmd: LaunchCommand): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    let settled = false;
    const child = spawn(cmd.command, cmd.args, {
      detached: true,
      stdio: 'ignore',
      shell: cmd.shell,
    });
    child.once('error', (err) => {
      if (settled) return;
      settled = true;
      reject(err);
    });
    child.once('spawn', () => {
      if (settled) return;
      settled = true;
      child.unref();
      resolve();
    });
  });
}

function resolveEditorCommand(env: Record<string, string | undefined>): string | undefined {
  for (const key of ['KIMI_CODE_EDITOR', 'VISUAL', 'EDITOR']) {
    const value = env[key];
    if (typeof value === 'string' && value.trim().length > 0) return value.trim();
  }
  return undefined;
}

function supportsLineTarget(command: string): boolean {
  const first = command.trim().split(/\s+/)[0] ?? '';
  return /(?:^|\/)(code|cursor|windsurf)(?:\.cmd|\.exe)?$/i.test(first);
}

function quoteShellArg(value: string, platform: NodeJS.Platform): string {
  if (platform === 'win32') return `"${value.replaceAll('"', '\\"')}"`;
  return `'${value.replaceAll("'", "'\\''")}'`;
}

File: packages/kimi-agent/src/server/file_launch.rs
//! Detached file/app launching for the `fs:open`, `fs:reveal` and `fs:open-in`
//! routes.
//!
//! Port of the retired `kap-server/src/lib/fileLaunch.ts` (MIT): resolve the
//! user's editor from the environment, otherwise fall back to the platform
//! opener (`open` / `cmd start` / `xdg-open`), and spawn it detached so the
//! server does not wait on the child.
//!
//! The previous Rust handlers discarded the resolved path and returned
//! `{"opened": true}` without launching anything — a fake success the web UI
//! surfaced as "file opened" while nothing happened. This module does the
//! actual launch and reports the error when it fails.

use std::path::Path;
use std::process::Command;

/// One app the `fs:open-in` route can target (v2 `OpenInAppId`).
pub const OPEN_IN_APP_IDS: [&str; 5] = ["finder", "cursor", "vscode", "iterm", "terminal"];

/// Editor env vars v2 honored, in precedence order.
const EDITOR_ENV_VARS: [&str; 3] = ["KIMI_CODE_EDITOR", "VISUAL", "EDITOR"];

/// Editors that accept a `path:line` target.
fn supports_line_target(editor: &str) -> bool {
    let lower = editor.to_ascii_lowercase();
    lower.contains("code")
        || lower.contains("cursor")
        || lower.contains("vim")
        || lower.contains("nvim")
        || lower.contains("subl")
}

/// Split a command line into executable and arguments, respecting quotes.
fn split_command_line(cmd: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut in_quotes = false;
    let mut quote_char = ' ';

    for c in cmd.chars() {
        match c {
            '"' | '\'' if in_quotes && c == quote_char => {
                in_quotes = false;
            }
            '"' | '\'' if !in_quotes => {
                in_quotes = true;
                quote_char = c;
            }
            c if c.is_whitespace() && !in_quotes => {
                if !current.is_empty() {
                    tokens.push(std::mem::take(&mut current));
                }
            }
            c => current.push(c),
        }
    }
    if !current.is_empty() {
        tokens.push(current);
    }
    tokens
}

/// Resolve the editor command from the environment, if one is set.
fn resolve_editor_command() -> Option<String> {
    for key in EDITOR_ENV_VARS {
        if let Ok(value) = std::env::var(key) {
            let trimmed = value.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }
    None
}

fn command_exists(command: &str) -> bool {
    #[cfg(windows)]
    let mut cmd = {
        let mut c = Command::new("where");
        c.arg(command);
        c
    };
    #[cfg(not(windows))]
    let mut cmd = {
        let mut c = Command::new("sh");
        c.arg("-c").arg(format!("command -v {command}"));
        c
    };
    cmd.stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

/// Explorer's select-item argument (v2 `explorerSelectArg`): `/select,<path>`
/// with backslashes, passed verbatim because explorer parses it itself.
fn explorer_select_arg(path: &Path) -> String {
    format!("/select,{}", path.display())
}

/// Launch a detached child. Returns the OS error string on failure.
fn launch_detached(program: &str, args: &[String]) -> Result<(), String> {
    #[cfg(test)]
    {
        // Avoid popping up actual GUI editor/file manager windows during test runs.
        let _ = (program, args);
        Ok(())
    }
    #[cfg(not(test))]
    {
        let mut command = Command::new(program);
        command
            .args(args)
            .stdin(std::process::Stdio::null())
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null());
        command
            .spawn()
            .map(|_child| ())
            .map_err(|e| format!("failed to launch {program}: {e}"))
    }
}

/// `fs:open`: open the file in the user's editor, or the platform default.
/// `line` is appended as `path:line` for editors that understand it.
pub fn open_file(absolute: &Path, line: Option<u64>) -> Result<(), String> {
    if let Some(editor) = resolve_editor_command() {
        let mut tokens = split_command_line(&editor);
        if !tokens.is_empty() {
            let program = tokens.remove(0);
            let target = if line.is_some() && supports_line_target(&program) {
                format!("{}:{}", absolute.display(), line.unwrap_or(0))
            } else {
                absolute.display().to_string()
            };
            tokens.push(target);
            return launch_detached(&program, &tokens);
        }
    }

    let path = absolute.display().to_string();
    if cfg!(target_os = "macos") {
        launch_detached("open", &[path])
    } else if cfg!(windows) {
        launch_detached("explorer.exe", &[path])
    } else {
        launch_detached("xdg-open", &[path])
    }
}

/// `fs:reveal`: select the file in the platform file manager.
pub fn reveal_file(absolute: &Path) -> Result<(), String> {
    let path = absolute.display().to_string();
    if cfg!(target_os = "macos") {
        launch_detached("open", &["-R".into(), path])
    } else if cfg!(windows) {
        let parent = absolute.parent().unwrap_or(absolute).display().to_string();
        if parent.is_empty() {
            launch_detached("explorer.exe", &[path])
        } else {
            launch_detached("explorer.exe", &[explorer_select_arg(absolute)])
        }
    } else {
        let dir = absolute.parent().unwrap_or(absolute).display().to_string();
        launch_detached("xdg-open", &[dir])
    }
}

/// Whether an `fs:open-in` app id is usable on this machine (v2
/// `getAvailableOpenInApps`).
pub fn is_open_in_app_available(app_id: &str) -> bool {
    match app_id {
        "finder" | "terminal" => cfg!(target_os = "macos"),
        "iterm" => {
            if !cfg!(target_os = "macos") {
                return false;
            }
            let home = std::env::var("HOME").unwrap_or_default();
            Path::new("/Applications/iTerm.app").exists()
                || Path::new(&format!("{home}/Applications/iTerm.app")).exists()
        }
        "vscode" => command_exists("code"),
        "cursor" => command_exists("cursor"),
        _ => false,
    }
}

/// `fs:open-in`: open the path in a specific app.
pub fn open_in_app(app_id: &str, absolute: &Path, line: Option<u64>) -> Result<(), String> {
    let path = absolute.display().to_string();
    match app_id {
        "vscode" | "cursor" => {
            let binary = if app_id == "vscode" { "code" } else { "cursor" };
            let args: Vec<String> = match line {
                Some(line) => vec!["-g".into(), format!("{}:{line}", absolute.display())],
                None => vec![path],
            };
            launch_detached(binary, &args)
        }
        "finder" => {
            if absolute.is_dir() {
                launch_detached("open", &[path])
            } else {
                launch_detached("open", &["-R".into(), path])
            }
        }
        "iterm" => launch_detached("open", &["-a".into(), "iTerm".into(), path]),
        "terminal" => launch_detached("open", &["-a".into(), "Terminal".into(), path]),
        other => Err(format!(
            "Unsupported app \"{other}\": expected one of {}",
            OPEN_IN_APP_IDS.join(", ")
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn open_in_app_rejects_unknown_id() {
        let err = open_in_app("photoshop", Path::new("/tmp/x"), None).unwrap_err();
        assert!(err.contains("Unsupported app"), "{err}");
        assert!(
            err.contains("vscode"),
            "the error lists the valid ids: {err}"
        );
    }

    #[test]
    fn open_missing_file_reports_launch_or_ok() {
        // Launching a platform opener with a nonexistent path either fails
        // (no opener installed) or succeeds in spawning. Both are acceptable;
        // the bug being guarded is returning success without spawning anything.
        let _ = open_file(Path::new("/definitely/not/here.txt"), None);
    }

    #[test]
    fn line_target_only_for_line_aware_editors() {
        assert!(supports_line_target("code"));
        assert!(supports_line_target("cursor"));
        assert!(supports_line_target("nvim"));
        assert!(!supports_line_target("notepad"));
    }

    #[test]
    fn test_split_command_line() {
        assert_eq!(split_command_line("code --wait"), vec!["code", "--wait"]);
        assert_eq!(
            split_command_line("\"C:\\Program Files\\app.exe\" -f \"file name\""),
            vec!["C:\\Program Files\\app.exe", "-f", "file name"]
        );
        assert_eq!(split_command_line("  notepad  "), vec!["notepad"]);
        assert!(split_command_line("   ").is_empty());
    }
}
